airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py

```diff
@@ -0,0 +1,394 @@
+            params["endTimestamp"] = end_ts
+            logger.info(f"Reading chunk from {ts} to {end_ts}")
+            yield from self.read(getter, params)
+
+    @property
+    def properties(self) -> Mapping[str, Any]:
+        if not self.entity:
+            return {}
+
+        props = {}
+        data = self._api.get(f"/properties/v2/{self.entity}/properties")
+        for row in data:
+            props[row["name"]] = {"type": row["type"]}
+
+        return props
+
+
+class CRMObjectsAPI(StreamAPI, ABC):
+    data_path = "results"
+
+    @property
+    @abstractmethod
+    def url(self):
+        """Endpoint URL"""
+
+    def __init__(self, include_archived_only=False, **kwargs):
+        super().__init__(**kwargs)
+        self._include_archived_only = include_archived_only
+
+    def list(self, fields) -> Iterable:
+        params = {
+            "archived": str(self._include_archived_only).lower(),
+        }
+        for record in self.read(partial(self._api.get, url=self.url), params):
+            yield record
+
+
+class CampaignsAPI(StreamAPI):
+    entity = "campaign"
+    more_key = "hasMore"
+    data_path = "campaigns"
+    limit = 500
+
+    def list(self, fields) -> Iterable:
+        url = "/email/public/v1/campaigns/by-id"
+        for row in self.read(getter=partial(self._api.get, url=url)):
+            record = self._api.get(f"/email/public/v1/campaigns/{row['id']}")
+            yield record
+
+
+class CompaniesAPI(CRMObjectsAPI):
+    entity = "company"
+    url = "/crm/v3/objects/companies"
+
+
+class ContactListsAPI(CRMObjectsAPI):
+    url = "/crm/v3/objects/contacts"
+
+
+class ContactsAPI(CRMObjectsAPI):
+    entity = "contact"
+    url = "/crm/v3/objects/contacts"
+
+
+class DealsAPI(CRMObjectsAPI):
+    entity = "deal"
+    url = "/crm/v3/objects/deals"
+
+
+class LineItemsAPI(CRMObjectsAPI):
+    entity = "line_item"
+    url = "/crm/v3/objects/line_items"
+
+
+class ProductsAPI(CRMObjectsAPI):
+    entity = "product"
+    url = "/crm/v3/objects/products"
+
+
+class QuotesAPI(CRMObjectsAPI):
+    entity = "quotes"
+    url = "/crm/v3/objects/quotes"
+
+
+class TicketsAPI(CRMObjectsAPI):
+    entity = "ticket"
+    url = "/crm/v3/objects/tickets"
+
+
+class ContactsByCompanyAPI(StreamAPI):
+    def list(self, fields) -> Iterable:
+        companies_api = CompaniesAPI(api=self._api, start_date=str(self._start_date))
+        for company in companies_api.list(fields={}):
+            yield from self._contacts_by_company(company["id"])
+
+    def _contacts_by_company(self, company_id):
+        url = "/companies/v2/companies/{pk}/vids".format(pk=company_id)
+        # FIXME: check if pagination is possible
+        params = {"count": 100}
+        path = "vids"
+        data = self._api.get(url, params)
+
+        if data.get(path) is None:
+            raise RuntimeError("Unexpected API response: {} not in {}".format(path, data.keys()))
+
+        for row in data[path]:
```

Comment:
is the point of this stream to be a join table between contacts and companies? can't we put this information in the contacts stream as a field?